src/bin/server.rs

```diff
@@ -62,17 +64,57 @@ fn register_user(user: Json<User>) -> Json<Value> {
 
 #[post("/users/login", format = "application/json", data="<user>")]
 fn login_user(user: Json<Value>) -> Json<Value> {
-    use schema::users;
+    use schema::auth_tokens;
 
     let connection = establish_connection();
 
     let t_username = user["username"].as_str().unwrap_or("");
     let t_password = encrypt_password(user["password"].as_str().unwrap_or(""));
 
-    let found_user = users.filter(username.eq(t_username).and(password.eq(t_password)))
-                          .load::<User>(&connection).ok();
+    let result = User::find_by_login(&connection, t_username, t_password.as_str());
+
+    match result {
+        Ok(user) => {
+            let rand_hash = gen_random_hash();
+            let expired_at = (epoch_now() as i64) + 30 * 24 * 60 * 60; // 30 days
+            let new_auth_token = AuthToken {
+                token: rand_hash,
+                expired_at: expired_at,
+                user_id: user.id,
+            };
+            let result: AuthToken = diesel::insert(&new_auth_token).into(auth_tokens::table).get_result(&connection)
+                .expect("Error creating auth token");
+
+            Json(json!({
+                "result": true,
+                "user": {
+                    "username": user.username,
+                    "email": user.email,
+                },
+                "token": result.token,
+            }))
+        },
+        Err(_) => {
+            Json(json!({
+                "result": false,
+            }))
+        }
+    }
+}
+
+#[get("/users/me")]
+fn get_user(auth: Auth) -> Json<Value> {
+    let connection = establish_connection();
+
+    let user = User::find(&connection, auth.user_id).unwrap();
```

Comment:
What if we could not find any user here? Panic? no no no no nooooo

Reply:
It's very unlikely we would be _panik_ here, since `Auth` middleware has already check whether user exists.

Reply:
How about user got deleted while login? Why don't we put `User` object inside `Auth`?